File: zig/hard/meeting_rooms_iii.zig
const std = @import("std");
const Allocator = std.mem.Allocator;
const PriorityQueue = std.PriorityQueue;

const Meeting = struct {
    start: i32,
    end: i32,
};

const BusyRoom = struct {
    finish_time: i32,
    room_id: i32,
};

fn compareMeetings(_: void, a: Meeting, b: Meeting) bool {
    return a.start < b.start;
}

fn compareFreeRooms(_: void, a: i32, b: i32) std.math.Order {
    return std.math.order(a, b);
}

fn compareBusyRooms(_: void, a: BusyRoom, b: BusyRoom) std.math.Order {
    if (a.finish_time != b.finish_time) {
        return std.math.order(a.finish_time, b.finish_time);
    }
    return std.math.order(a.room_id, b.room_id);
}

pub fn mostBooked(allocator: Allocator, n: i32, meetings: []Meeting) !i32 {
    // Sort meetings by start time
    std.mem.sort(Meeting, meetings, {}, compareMeetings);
    
    // Initialize free rooms heap
    var free_rooms = PriorityQueue(i32, void, compareFreeRooms).init(allocator, {});
    defer free_rooms.deinit();
    
    var i: i32 = 0;
    while (i < n) : (i += 1) {
        try free_rooms.add(i);
    }
    
    // Initialize busy rooms heap
    var busy_rooms = PriorityQueue(BusyRoom, void, compareBusyRooms).init(allocator, {});
    defer busy_rooms.deinit();
    
    // Initialize booking count
    const booking_count = try allocator.alloc(i32, @intCast(n));
    defer allocator.free(booking_count);
    @memset(booking_count, 0);
    
    // Process each meeting
    for (meetings) |meeting| {
        const start = meeting.start;
        const end = meeting.end;
        const duration = end - start;
        
        // Release rooms that finished before current meeting starts
        while (busy_rooms.peek()) |top| {
            if (top.finish_time <= start) {
                const busy = busy_rooms.remove();
                try free_rooms.add(busy.room_id);
            } else {
                break;
            }
        }
        
        var room_id: i32 = undefined;
        
        if (free_rooms.peek()) |_| {
            // Assign to free room with lowest number
            room_id = free_rooms.remove();
            try busy_rooms.add(.{ .finish_time = end, .room_id = room_id });
        } else {
            // Delay meeting to earliest finishing room
            const busy = busy_rooms.remove();
            room_id = busy.room_id;
            try busy_rooms.add(.{
                .finish_time = busy.finish_time + duration,
                .room_id = room_id,
            });
        }
        
        booking_count[@intCast(room_id)] += 1;
    }
    
    // Find room with maximum bookings
    var max_bookings: i32 = 0;
    var result: i32 = 0;
    
    i = 0;
    while (i < n) : (i += 1) {
        if (booking_count[@intCast(i)] > max_bookings) {
            max_bookings = booking_count[@intCast(i)];
            result = i;
        }
    }
    
    return result;
}

File: typescript/hard/meeting-rooms-iii.ts
// Min-heap implementation for TypeScript
class MinHeap<T> {
    private heap: T[] = [];

    constructor(private compareFn: (a: T, b: T) => number) {}

    push(item: T): void {
        this.heap.push(item);
        this.bubbleUp(this.heap.length - 1);
    }

    pop(): T | undefined {
        if (this.heap.length === 0) return undefined;
        if (this.heap.length === 1) return this.heap.pop();

        const root = this.heap[0];
        this.heap[0] = this.heap.pop()!;
        this.bubbleDown(0);
        return root;
    }

    peek(): T | undefined {
        return this.heap[0];
    }

    get size(): number {
        return this.heap.length;
    }

    private bubbleUp(index: number): void {
        while (index > 0) {
            const parentIndex = Math.floor((index - 1) / 2);
            if (this.compareFn(this.heap[index], this.heap[parentIndex]) >= 0)
                break;

            [this.heap[index], this.heap[parentIndex]] = [
                this.heap[parentIndex],
                this.heap[index],
            ];
            index = parentIndex;
        }
    }

    private bubbleDown(index: number): void {
        while (true) {
            let smallest = index;
            const leftChild = 2 * index + 1;
            const rightChild = 2 * index + 2;

            if (
                leftChild < this.heap.length &&
                this.compareFn(this.heap[leftChild], this.heap[smallest]) < 0
            ) {
                smallest = leftChild;
            }

            if (
                rightChild < this.heap.length &&
                this.compareFn(this.heap[rightChild], this.heap[smallest]) < 0
            ) {
                smallest = rightChild;
            }

            if (smallest === index) break;

            [this.heap[index], this.heap[smallest]] = [
                this.heap[smallest],
                this.heap[index],
            ];
            index = smallest;
        }
    }
}

interface BusyRoom {
    finishTime: number;
    roomId: number;
}

function mostBooked(n: number, meetings: number[][]): number {
    // Sort meetings by start time
    meetings.sort((a, b) => a[0] - b[0]);

    // Initialize free rooms heap (min-heap by room number)
    const freeRooms = new MinHeap<number>((a, b) => a - b);
    for (let i = 0; i < n; i++) {
        freeRooms.push(i);
    }

    // Initialize busy rooms heap (min-heap by finish time, then by room id)
    const busyRooms = new MinHeap<BusyRoom>((a, b) => {
        if (a.finishTime !== b.finishTime) {
            return a.finishTime - b.finishTime;
        }
        return a.roomId - b.roomId;
    });

    // Initialize booking count
    const bookingCount = new Array(n).fill(0);

    // Process each meeting
    for (const [start, end] of meetings) {
        const duration = end - start;

        // Release rooms that finished before current meeting starts
        while (busyRooms.size > 0 && busyRooms.peek()!.finishTime <= start) {
            const { roomId } = busyRooms.pop()!;
            freeRooms.push(roomId);
        }

        let roomId: number;

        if (freeRooms.size > 0) {
            // Assign to free room with lowest number
            roomId = freeRooms.pop()!;
            busyRooms.push({ finishTime: end, roomId });
        } else {
            // Delay meeting to earliest finishing room
            const { finishTime, roomId: busyRoomId } = busyRooms.pop()!;
            roomId = busyRoomId;
            busyRooms.push({
                finishTime: finishTime + duration,
                roomId,
            });
        }

        bookingCount[roomId]++;
    }

    // Find room with maximum bookings (return lowest number if tie)
    let maxBookings = 0;
    let result = 0;

    for (let i = 0; i < n; i++) {
        if (bookingCount[i] > maxBookings) {
            maxBookings = bookingCount[i];
            result = i;
        }
    }

    return result;
}
